File: app/src/main/java/com/numero/sojodia/presenter/BusSchedulePresenter.java
package com.numero.sojodia.presenter;

import com.numero.sojodia.contract.BusScheduleContract;
import com.numero.sojodia.model.BusDataFile;
import com.numero.sojodia.model.BusTime;
import com.numero.sojodia.model.Reciprocate;
import com.numero.sojodia.model.Time;
import com.numero.sojodia.repository.BusDataRepository;
import com.numero.sojodia.util.TimeUtil;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.Single;

public class BusSchedulePresenter implements BusScheduleContract.Presenter {

    private static final int NO_BUS_POSITION = -1;

    private final BusDataRepository busDataRepository;

    private final BusScheduleContract.View view;
    private final Reciprocate reciprocate;
    private int week = 0;

    private Integer tkBusPosition;
    private Integer tndBusPosition;

    private List<BusTime> tkBusTimeListGoing = new ArrayList<>();
    private List<BusTime> tkBusTimeListReturn = new ArrayList<>();
    private List<BusTime> tndBusTimeListGoing = new ArrayList<>();
    private List<BusTime> tndBusTimeListReturn = new ArrayList<>();

    public BusSchedulePresenter(BusScheduleContract.View view, BusDataRepository busDataRepository, Reciprocate reciprocate) {
        this.view = view;
        this.busDataRepository = busDataRepository;
        this.reciprocate = reciprocate;

        this.view.setPresenter(this);

        initBusDataList();
    }

    @Override
    public void subscribe() {
    }

    @Override
    public void unSubscribe() {
    }

    @Override
    public void onTimeChanged(int week) {
        this.week = week;
        view.showTkBusTimeList(getTkTimeList());
        view.showTndBusTimeList(getTndTimeList());
        view.selectTkCurrentBusPosition(getTkBusPosition());
        view.selectTndCurrentBusPosition(getTndBusPosition());
        // TODO if で position = NO_BUSかで分岐させる
        view.startTkCountDown(getTkTimeList().get(getTkBusPosition()));
        view.startTndCountDown(getTndTimeList().get(getTndBusPosition()));
        if (canPreviewTkTime()) {
            view.showTkPreviewButton();
        } else {
            view.hideTkPreviewButton();
        }
        if (canPreviewTndTime()) {
            view.showTndPreviewButton();
        } else {
            view.hideTndPreviewButton();
        }
        if (isNoTkBus()) {
            view.showTkNoBusLayout();
        } else {
            view.hideTkNoBusLayout();
        }
        if (isNoTndBus()) {
            view.showTndNoBusLayout();
        } else {
            view.hideTndNoBusLayout();
        }

    }

    @Override
    public void nextTkBus() {
        // getTkBusPosition でもいい?
        if (tkBusPosition == NO_BUS_POSITION) {
            return;
        }
        tkBusPosition += 1;
        if (tkBusPosition == getTkTimeList().size()) {
            tkBusPosition = NO_BUS_POSITION;
        }
        if (tkBusPosition >= getTkTimeList().size() - 1) {
            view.hideTkNextButton();
        } else {
            view.showTkNextButton();
        }
        if (canPreviewTkTime()) {
            view.showTkPreviewButton();
        } else {
            view.hideTkPreviewButton();
        }
        view.selectTkCurrentBusPosition(tkBusPosition);
        view.startTkCountDown(getTkTimeList().get(getTkBusPosition()));
    }

    @Override
    public void previewTkBus() {
        tkBusPosition -= 1;
        view.selectTkCurrentBusPosition(tkBusPosition);
        view.startTkCountDown(getTkTimeList().get(getTkBusPosition()));
        if (canPreviewTkTime()) {
            view.showTkPreviewButton();
        } else {
            view.hideTkPreviewButton();
        }
    }

    @Override
    public void nextTndBus() {
        // FIXME
        if (tndBusPosition == NO_BUS_POSITION) {
            return;
        }
        tndBusPosition += 1;
        if (tndBusPosition == getTndTimeList().size()) {
            tndBusPosition = NO_BUS_POSITION;
        }
        if (tndBusPosition >= getTndTimeList().size() - 1) {
            view.hideTndNextButton();
        } else {
            view.showTndNextButton();
        }
        if (canPreviewTndTime()) {
            view.showTndPreviewButton();
        } else {
            view.hideTndPreviewButton();
        }
        view.selectTndCurrentBusPosition(getTndBusPosition());
        view.startTndCountDown(getTndTimeList().get(getTndBusPosition()));
    }

    @Override
    public void previewTndBus() {
        tndBusPosition -= 1;
        view.selectTndCurrentBusPosition(getTndBusPosition());
        view.startTndCountDown(getTndTimeList().get(getTndBusPosition()));
        if (canPreviewTndTime()) {
            view.showTndPreviewButton();
        } else {
            view.hideTndPreviewButton();
        }
    }

    public void setWeek(int week) {
        this.week = week;
    }

    public int getTkBusPosition() {
        if (tkBusPosition == null) {
            tkBusPosition = findBusPosition(getTkTimeList());
        }
        return tkBusPosition;
    }

    public int getTndBusPosition() {
        if (tndBusPosition == null) {
            tndBusPosition = findBusPosition(getTndTimeList());
        }
        return tndBusPosition;
    }

    public List<BusTime> getTkTimeList() {
        // TODO キャッシュさせる
        return getTkBusTimeListObservable().blockingGet();
    }

    public List<BusTime> getTndTimeList() {
        // TODO キャッシュさせる
        return getTndBusTimeListObservable().blockingGet();
    }

    private void initBusDataList() {
        tkBusTimeListGoing = busDataRepository.loadBusData(BusDataFile.TK_TO_KUTC).blockingFirst();
        tkBusTimeListReturn = busDataRepository.loadBusData(BusDataFile.KUTC_TO_TK).blockingFirst();
        tndBusTimeListGoing = busDataRepository.loadBusData(BusDataFile.TND_TO_KUTC).blockingFirst();
        tndBusTimeListReturn = busDataRepository.loadBusData(BusDataFile.KUTC_TO_TND).blockingFirst();
    }

    private int findBusPosition(List<BusTime> busTimeList) {
        int position;
        Time time = new Time();
        for (position = 0; position < busTimeList.size(); position++) {
            time.setTime(busTimeList.get(position).hour, busTimeList.get(position).min, 0);
            if (TimeUtil.isOverTime(new Time(), time)) {
                return position;
            }
        }
        return NO_BUS_POSITION;
    }

    private Single<List<BusTime>> getTkBusTimeListObservable() {
        switch (reciprocate) {
            case GOING:
                return Observable.fromIterable(tkBusTimeListGoing)
                        .filter(busTime -> busTime.week == week)
                        .toList();
            case RETURN:
            default:
                return Observable.fromIterable(tkBusTimeListReturn)
                        .filter(busTime -> busTime.week == week)
                        .toList();
        }
    }

    private Single<List<BusTime>> getTndBusTimeListObservable() {
        switch (reciprocate) {
            case GOING:
                return Observable.fromIterable(tndBusTimeListGoing)
                        .filter(busTime -> busTime.week == week)
                        .toList();
            case RETURN:
            default:
                return Observable.fromIterable(tndBusTimeListReturn)
                        .filter(busTime -> busTime.week == week)
                        .toList();
        }
    }

    private boolean isNoTkBus() {
        return getTkBusPosition() == NO_BUS_POSITION;
    }

    private boolean isNoTndBus() {
        return getTndBusPosition() == NO_BUS_POSITION;
    }

    private boolean canPreviewTkTime() {
        if (tkBusPosition == NO_BUS_POSITION || tkBusPosition == 0) {
            return false;
        }
        BusTime busTime = getTkTimeList().get(tkBusPosition - 1);
        Time time = new Time();
        time.setTime(busTime.hour, busTime.min, 0);
        return TimeUtil.isOverTime(new Time(), time);
    }

    private boolean canNextTkTime() {
        return tkBusPosition != NO_BUS_POSITION && !(tkBusPosition + 1 >= getTkTimeList().size());
    }

    private boolean canPreviewTndTime() {
        if (tndBusPosition == NO_BUS_POSITION || tndBusPosition == 0) {
            return false;
        }
        BusTime busTime = getTndTimeList().get(tndBusPosition - 1);
        Time time = new Time();
        time.setTime(busTime.hour, busTime.min, 0);
        return TimeUtil.isOverTime(new Time(), time);
    }

    public boolean canNextTndTime() {
        return tndBusPosition != NO_BUS_POSITION && !(tndBusPosition + 1 >= getTndTimeList().size());
    }
}

File: app/src/main/kotlin/com/numero/sojodia/presenter/BusSchedulePresenter.kt
package com.numero.sojodia.presenter

import com.numero.sojodia.contract.BusScheduleContract
import com.numero.sojodia.extension.isOverTime
import com.numero.sojodia.model.*
import com.numero.sojodia.repository.IBusDataRepository

class BusSchedulePresenter(private val view: BusScheduleContract.View,
                           private val busDataRepository: IBusDataRepository,
                           private val reciprocate: Reciprocate) : BusScheduleContract.Presenter {
    private lateinit var week: Week

    private var tkBusPosition = 0
    private var tndBusPosition = 0

    private var tkTimeList: MutableList<BusTime> = mutableListOf()
        get() {
            return if (field.isEmpty()) {
                when (reciprocate) {
                    Reciprocate.GOING -> busDataRepository.tkBusTimeListGoing
                            .filter { busTime -> busTime.week == week }.toMutableList()
                            .apply {
                                field = this
                            }
                    Reciprocate.RETURN -> busDataRepository.tkBusTimeListReturn
                            .filter { busTime -> busTime.week == week }.toMutableList()
                            .apply {
                                field = this
                            }
                }
            } else {
                field
            }
        }

    private var tndTimeList: MutableList<BusTime> = mutableListOf()
        get() {
            return if (field.isEmpty()) {
                when (reciprocate) {
                    Reciprocate.GOING -> busDataRepository.tndBusTimeListGoing
                            .filter { busTime -> busTime.week == week }.toMutableList()
                            .apply {
                                field = this
                            }
                    Reciprocate.RETURN -> busDataRepository.tndBusTimeListReturn
                            .filter { busTime -> busTime.week == week }.toMutableList()
                            .apply {
                                field = this
                            }
                }
            } else {
                field
            }
        }

    init {
        this.view.setPresenter(this)
    }

    override fun subscribe() {}

    override fun unSubscribe() {}

    override fun onTimeChanged(week: Week) {
        this.week = week
        tkTimeList.clear()
        tndTimeList.clear()
        view.showTkBusTimeList(tkTimeList)
        view.showTndBusTimeList(tndTimeList)

        tkBusPosition = findBusPosition(tkTimeList)
        tndBusPosition = findBusPosition(tndTimeList)

        if (tkBusPosition == NO_BUS_POSITION) {
            view.showTkNoBusLayout()
            view.hideTkNextButton()
            view.hideTkPreviewButton()
        } else {
            view.hideTkNoBusLayout()
            view.selectTkCurrentBusPosition(tkBusPosition)
            view.startTkCountDown(tkTimeList[tkBusPosition])
            if (tkBusPosition >= tkTimeList.size - 1) {
                view.hideTkNextButton()
            } else {
                view.showTkNextButton()
            }
            if (canPreviewTkTime()) {
                view.showTkPreviewButton()
            } else {
                view.hideTkPreviewButton()
            }
        }
        if (tndBusPosition == NO_BUS_POSITION) {
            view.showTndNoBusLayout()
            view.hideTndNextButton()
            view.hideTndPreviewButton()
        } else {
            view.hideTndNoBusLayout()
            view.selectTndCurrentBusPosition(tndBusPosition)
            view.startTndCountDown(tndTimeList[tndBusPosition])
            if (tndBusPosition >= tndTimeList.size - 1) {
                view.hideTndNextButton()
            } else {
                view.showTndNextButton()
            }
            if (canPreviewTndTime()) {
                view.showTndPreviewButton()
            } else {
                view.hideTndPreviewButton()
            }
        }
    }

    override fun nextTkBus() {
        if (tkBusPosition == tkTimeList.size - 1) {
            tkBusPosition = NO_BUS_POSITION
        }
        if (tkBusPosition == NO_BUS_POSITION) {
            view.hideTkNextButton()
            view.hideTkPreviewButton()
            view.showTkNoBusLayout()
            return
        }
        tkBusPosition += 1
        if (tkBusPosition >= tkTimeList.size - 1) {
            view.hideTkNextButton()
        } else {
            view.showTkNextButton()
        }
        if (canPreviewTkTime()) {
            view.showTkPreviewButton()
        } else {
            view.hideTkPreviewButton()
        }
        view.selectTkCurrentBusPosition(tkBusPosition)
        view.startTkCountDown(tkTimeList[tkBusPosition])
    }

    override fun previewTkBus() {
        tkBusPosition -= 1
        view.selectTkCurrentBusPosition(tkBusPosition)
        view.startTkCountDown(tkTimeList[tkBusPosition])
        if (canPreviewTkTime()) {
            view.showTkPreviewButton()
        } else {
            view.hideTkPreviewButton()
        }
        if (tkBusPosition >= tkTimeList.size - 1) {
            view.hideTkNextButton()
        } else {
            view.showTkNextButton()
        }
    }

    override fun nextTndBus() {
        if (tndBusPosition == tndTimeList.size - 1) {
            tndBusPosition = NO_BUS_POSITION
        }
        if (tndBusPosition == NO_BUS_POSITION) {
            view.hideTndNextButton()
            view.hideTndPreviewButton()
            view.showTndNoBusLayout()
            return
        }
        tndBusPosition += 1
        if (tndBusPosition >= tndTimeList.size - 1) {
            view.hideTndNextButton()
        } else {
            view.showTndNextButton()
        }
        if (canPreviewTndTime()) {
            view.showTndPreviewButton()
        } else {
            view.hideTndPreviewButton()
        }
        view.selectTndCurrentBusPosition(tndBusPosition)
        view.startTndCountDown(tndTimeList[tndBusPosition])
    }

    override fun previewTndBus() {
        tndBusPosition -= 1
        view.selectTndCurrentBusPosition(tndBusPosition)
        view.startTndCountDown(tndTimeList[tndBusPosition])
        if (canPreviewTndTime()) {
            view.showTndPreviewButton()
        } else {
            view.hideTndPreviewButton()
        }
        if (tndBusPosition >= tndTimeList.size - 1) {
            view.hideTndNextButton()
        } else {
            view.showTndNextButton()
        }
    }

    override fun showTimeTableDialog(route: Route) {
        view.showTimeTableDialog(route, reciprocate)
    }

    private fun findBusPosition(busTimeList: List<BusTime>): Int {
        busTimeList.forEachIndexed { index, busTime ->
            if (busTime.time.isOverTime(Time())) {
                return index
            }
        }
        return NO_BUS_POSITION
    }

    private fun canPreviewTkTime(): Boolean {
        if (tkBusPosition == NO_BUS_POSITION || tkBusPosition == 0) {
            return false
        }
        val busTime = tkTimeList[tkBusPosition - 1]
        return busTime.time.isOverTime(Time())
    }

    private fun canPreviewTndTime(): Boolean {
        if (tndBusPosition == NO_BUS_POSITION || tndBusPosition == 0) {
            return false
        }
        val busTime = tndTimeList[tndBusPosition - 1]
        return busTime.time.isOverTime(Time())
    }

    companion object {
        private const val NO_BUS_POSITION = -1
    }
}
